Add federation support

tuxmachine opened the pull request:
# PR Checklist
Please check if your PR fulfills the following requirements:

- [X] The commit message follows our guidelines: https://github.com/nestjs/nest/blob/master/CONTRIBUTING.md
- [X] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->
```
[ ] Bugfix
[X] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[X] Build related changes
[ ] CI related changes
[ ] Other... Please describe:
```

## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number:
#299
#288


## What is the new behavior?

Support for Apollo Federation

## Does this PR introduce a breaking change?
```
[ ] Yes
[X] No
```

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

This is a follow-up PR to #301 by @marcus-sa . Many thanks to his original implementation, I want to use it and had some spare time, so I implemented the unit tests and will start the documentation tomorrow. Don't know the correct etiquette for this sort of situation, opening up a new PR seemed easiest.

tuxmachine commented:
I could still use some help for one more test case I'd like to cover. Marcus provided a string token to register a `buildService` as a nest injectable, which will be passed to Apollo Gateway. However, as I'm relatively new to Nest, I couldn't figure out how to fill this injectable :shrug: Could really use some suggestions here :) 

```ts
// graphql-gateway.module.ts
@Module({})
export class GraphQLGatewayModule implements OnModuleInit {
  private apolloServer: ApolloServerBase;

  constructor(
    @Optional()
    private readonly httpAdapterHost: HttpAdapterHost,

    @Optional()
    @Inject(GRAPHQL_GATEWAY_BUILD_SERVICE) // <-- This thing here

    private readonly buildService: GatewayBuildService,
    @Inject(GRAPHQL_GATEWAY_MODULE_OPTIONS)
    private readonly options: GatewayModuleOptions,
  ) {}
```

I figured something like this would work, but sadly no luck

```ts
class BuildService extends RemoteGraphQLDataSource {
  willSendRequest() {
    console.log('RECEIVED!!');
  }
}

@Module({
  imports: [
    GraphQLGatewayModule.forRoot(/* config */),
  ],
  providers: [{
    provide: GRAPHQL_GATEWAY_BUILD_SERVICE,
    useClass: BuildService,
  }],
})
export class AppModule {}
```

marcus-sa commented:
Btw you should upgrade the Apollo dependencies as there's a breaking change in the gateway class options.

The way I initially wanted to implement the BuildService was an injectable that works like an async options factory basically.

HINT: The reason why the provider is unavailable in the module context, is because:
1. The imported module is in a different scope
2. Imported modules gets initialized before providers

tuxmachine commented:
I rebased your branch on master where Apollo is already updated but I forgot to check the changelog. Will do that tomorrow, thanks for the heads up!

Figured if was something like that, tried injecting it into a dynamic module using the `forRootAsync` but couldn't get it working. Will give it another shot tomorrow.

tuxmachine commented:
* Added test for custom buildservice
* Updated `@apollo/{gateway,federation}` dependencies
* Started docs on tuxmachine/docs.nestjs.com 

@kamilmysliwiec Ready for a review 👍

yeuem1vannam commented:
Does this PR also provide a way to share the context across services?
References: https://www.apollographql.com/docs/apollo-server/federation/implementing/#sharing-context-across-services

tuxmachine commented:
Yes it does, but maybe the docs need to be clearer om that.

https://deploy-preview-802--docs-nestjs.netlify.com/graphql/federation#build-service

tuxmachine commented on lib/graphql-gateway.module.ts:
```diff
@@ -0,0 +1,174 @@
+      this.options.server || {};
+    const app = this.httpAdapterHost.httpAdapter.getInstance();
+
+    const apolloServer = new ApolloServer(apolloOptions);
```
You're right, I think I addressed the issue. with the latest push

tuxmachine commented:
If there's anything more that I can do regarding this PR, please let me know :) happy to spend some more time on it.

ZFLloyd commented:
@tuxmachine Have you tested [Custom scalars and Enum internal values features](https://www.apollographql.com/docs/graphql-tools/scalars/) from ApolloServer with this  ?

I tried federation using @marcus-sa branch you started working from I believe and passing scalars/enums in the resolvers argument when importing GraphQLFederationModule did not work. For example with the GraphQLJSON Scalar from graphql-type-json package:

```
GraphQLFederationModule.forRoot({
      typePaths: [...GQLTypesPath, './**/*.type.gql'],
      resolvers: { JSON: GraphQLJSON },
      context,
    })
```
It seems the JSON scalar is basically ignored, and any kind of input can go through.

marcus-sa commented:
@ZFLloyd nope, I think I never got to that point.

tuxmachine commented:
@ZFLloyd Thanks for highlighting a missing piece. My real-world experience with both nest and graphql are limited, so the feedback is really valuable. I'll see what I can do this week.

yeuem1vannam commented:
@tuxmachine @marcus-sa 
Thank you so much for the implementation of this feature, this is super useful for me right now.

However, when working with Apollo Graph Manager, this line fire error
```ts
// lib/graphql-gateway.module.ts:L107
const { schema, executor } = await gateway.load();
```
<details>
<summary>Error Log</summary>

```log
[WARN] Tue Nov 05 2019 14:35:44 GMT+0000 (Coordinated Universal Time) apollo-gateway: Error checking for schema updates. Falling back to existing schema. Error: When `serviceList` is not set, an Apollo Engine configuration must be provided. See https://www.apollographql.com/docs/apollo-server/federation/managed-federation/ for more information.
    at ApolloGateway.<anonymous> (/node_modules/@apollo/gateway/dist/index.js:271:23)
    at Generator.next (<anonymous>)
    at /node_modules/@apollo/gateway/dist/index.js:8:71
    at new Promise (<anonymous>)
    at __awaiter (/node_modules/@apollo/gateway/dist/index.js:4:12)
    at ApolloGateway.loadServiceDefinitions (/node_modules/@apollo/gateway/dist/index.js:260:16)
    at ApolloGateway.<anonymous> (/node_modules/@apollo/gateway/dist/index.js:162:37)
    at Generator.next (<anonymous>)
    at /node_modules/@apollo/gateway/dist/index.js:8:71
    at new Promise (<anonymous>)
(node:1515) UnhandledPromiseRejectionWarning: Error: Apollo Server requires either an existing schema, modules or typeDefs
    at ApolloServer.initSchema (/node_modules/apollo-server-core/dist/ApolloServer.js:240:23)
    at new ApolloServerBase (/node_modules/apollo-server-core/dist/ApolloServer.js:202:30)
    at new ApolloServer (/node_modules/apollo-server-express/dist/ApolloServer.js:59:9)
    at GraphQLGatewayModule.registerExpress (/workspace/packages/nestjs-graphql-pull-455/dist/graphql-gateway.module.js:99:30)
    at GraphQLGatewayModule.registerGqlServer (/workspace/packages/nestjs-graphql-pull-455/dist/graphql-gateway.module.js:86:18)
    at GraphQLGatewayModule.<anonymous> (/workspace/packages/nestjs-graphql-pull-455/dist/graphql-gateway.module.js:76:18)
    at Generator.next (<anonymous>)
    at fulfilled (/node_modules/tslib/tslib.js:107:62)
    at process._tickCallback (internal/process/next_tick.js:68:7)
    at Function.Module.runMain (internal/modules/cjs/loader.js:834:11)
(node:1515) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 2)
(node:1515) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

</details>

so I changed the code as below to make it works.
```ts
// lib/graphql-gateway.module.ts:L102-L108
const gateway = new ApolloGateway({
  ...gatewayOpts,
  buildService,
});

// const { schema, executor } = await gateway.load(); // Skip the load at this point.
this.registerGqlServer({ ...serverOpts, gateway, subscriptions: false });
```
By doing so, both the production (use Apollo Graph Manager) and development (use `serviceList` as usual) work as expected.

References:
- https://www.apollographql.com/docs/graph-manager/federation/#registering-federated-services
- https://www.apollographql.com/docs/graph-manager/federation/#connecting-apollo-server-to-graph-manager

tuxmachine commented:
@yeuem1vannam Thanks for testing with Graph Manager! Your fix works with the existing tests so I added it to the PR.

tuxmachine commented:
@ZFLloyd I believe I've fixed it. Could you verify?

renepardon commented:
One question left: will this implementation work with the `type-graphql` package to support the code first development?

marcus-sa commented:
@renepardon nope, `type-graphql` does not support custom directives.

renepardon commented:
@marcus-sa so in this case i should not focus on the code first approach rather then writing the schema by myself? Good to know before I start adding to much decorators to my classes :)

renepardon commented:
Should we create a new issue which stays open until https://github.com/MichalLytek/type-graphql/issues/77 is released?

tuxmachine commented:
It looks like `type-graphql` added support for federation directives (not *all* directives) just this weekend :tada: https://github.com/MichalLytek/type-graphql/issues/351#issuecomment-549156066 

So yeah, this PR fixes the schema-first approach. There needs to be extra work done for the code-first approach but it appears to no longer be blocking

tuxmachine commented:
@kamilmysliwiec I'm assuming you'd like to wait with this feature until it can support the code-first approach? 

Other than that, anything else you'd like to see?

lgabeskiria commented:
@kamilmysliwiec do you have any ETA for this feature to be released?

eole1712 commented on lib/graphql-federation.module.ts:
```diff
@@ -0,0 +1,204 @@
+
+    const { typePaths } = this.options;
+    const typeDefs = await this.graphqlTypesLoader.getTypesFromPaths(typePaths);
+    const mergedTypeDefs = extend(typeDefs, this.options.typeDefs);
```
Hello @tuxmachine
Why do you use here getTypesFromPaths instead of mergeTypesByPaths, as it is the case in graphql.module.ts?

eole1712 replied:
I think because of this, when we try to create an app with multiple resolvers and gql files, it creates errors of the type `GraphQLError: There can be only one type named "Query"` since it no longer merges these types together.
This is breaking compared to the behavior of graphql.module.ts and the Nest.js documentation where we are allowed to declare the same Query and Mutation types several times

tuxmachine replied:
Are you using `extend type Query` ? Without the `extend` I get the same error. 

I assumed @marcus-sa had a good reason not to use the `mergeTypesByPaths`, I'll look into it.

marcus-sa replied:
You need to use `extend type` otherwise it won't work (this is also documented in Apollo Federation & GraphQL)

I remember I had a good reason for not using it, but it's been so long that I've forgot.

tuxmachine replied:
Thanks for pitching in :) I'll have a look 

The `mergeTypesByPaths` uses an external library, so I can imagine it didn't support federation back then

tuxmachine replied:
`mergeTypesByPaths` resolves to nearly an identical schema (just adds a `schema` property) so the issue seems to have been resolved.

eole1712 replied:
seems great, thanks @tuxmachine 💚

eole1712 commented on lib/interfaces/gql-gateway-module-options.interface.ts:
```diff
@@ -0,0 +1,30 @@
+
+export interface GatewayModuleOptions {
+  gateway?: GatewayConfig;
+  server?: Pick<
```
@tuxmachine you pick here only few fields that we can pass to apollo when we create our server. 
Is there any particular reason for that? 😕 

This prevents us, for example, from initializing and configuring Apollo Graph Manager with the field `engine`, which can be quite annoying for some implementations with a distributed graph.

tuxmachine replied:
Again, I followed the original code of @marcus-sa that did a `Pick<GqlModuleOptions>`, maybe he can chime in with his original intention? I assume it's because the code doesn't support all the options, nor do they make sense for the gateway (e.g. `autoSchemaFile`) 

Are there any other fields you are missing?

tuxmachine replied:
I inverted the Pick so the engine prop becomes available

eole1712 commented:
@tuxmachine seems great to me, I tested on a very large project (9 services, 72 modules), it works perfectly now. Thanks for the changes ! ❤️

juicycleff commented:
@tuxmachine for code-first approach, please you could see how I did it here https://github.com/juicycleff/nestjs-graphql-gateway. It's working good should you want to learn from it, it's quite basic now. Sorry I wish to contribute, but currently overwhelmed by other things

juicycleff commented:
Once it's fully integrated, I can retire the library

tuxmachine commented:
@juicycleff that's awesome! I'll see if I can have a closer look this weekend to integrate it

MikeyUchiha commented:
@tuxmachine I was wondering if you have a time frame on this release? :)

tuxmachine commented:
@mikeyuchiha I got some ways in with the code-first approach. Need to test it some more and write tests for it. Shouldn't take more than 2 weeks.

As for release, you'll have to ask the maintainer @kamilmysliwiec but as the maintainer of the entire nestjs framework, he's rather busy

tuxmachine commented:
@MikeyUchiha @juicycleff I added the support for code-first. Will update the docs somewhere this week. However I disabled it for now until `type-graphql@2.0.0` lands with support for Directives. I doubt Kamil wants a `beta` dependency. 

I've got one question for the original contributors to this module. What was the purpose of the `GraphQLFactory#overrideOrExtendResolvers` because applying the same logic to the federated schema will break it. (I know @juicycleff uses it, but you're merging the same schema with itself, so won't make any changes)

MikeyUchiha commented:
@tuxmachine great job! hopefully they come out with an updated version soon.

antogyn commented on lib/graphql-federation.factory.ts:
```diff
@@ -0,0 +1,100 @@
+      // Enable support when Directive support in type-graphql goes stable
+      throw new Error('Code-first not supported yet');
+      schema = await this.generateSchema(options);
+    } else if (!isEmpty(options.typeDefs)) {
```
I think it's supposed to be:

```ts
} else if (!isEmpty(options.schema)) {
```

tuxmachine replied:
Nope, it follows the design from the original `graphql.factory.ts`. If you don't pass an array of typeDefs that will be used to construct a schema, then you are required to provide your own schema.

Davide-Gheri replied:
so shouldn't that be:
```typescript
else if (isEmpty(options.typeDefs)) {
```
as it is in https://github.com/nestjs/graphql/blob/master/lib/graphql.factory.ts#L77 ?

Now if you try to instantiate passing typePaths in options, it will throw an error because options.schema is undefined, if you pass an existing schema with options.schema, it will throw because it try to `buildSchemaFromTypeDefs` where typeDefs is `null`

tuxmachine replied:
😅 you are absolutely correct. Thank you very much!

I forgot to disable the unittest filter when I last pushed this branch, it was still set only running type-graphql tests.

tuxmachine commented:
@kamilmysliwiec I understand this turned into quite a big PR that'll be laborious te review... Any idea about an ETA?

marcus-sa commented:
I'll do some testing ik the weekend @tuxmachine

MikeyUchiha commented:
@marcus-sa thank you!

LucienLee commented:
Any progress on this PR? Is anything I can help with this feature?

tuxmachine commented:
@LucienLee Not really, it's finished as far as I know. There's just a complete lack of involvement from the primary maintainer. I've used this module for the past 2 months now without much issues, granted it hasn't seen a production load yet but I'm not expecting difficulties.

peshkov3 commented:
When this PR will be resolved?

XBeg9 commented:
Waiting for this PR, any help needed?

lgabeskiria commented:
@tuxmachine can you release this feature as a separate package? we're all waiting for this feature and there is no involvement from @kamilmysliwiec

kamilmysliwiec commented:
@lgabeskiria I have tons of other packages to maintain.

This PR is quite big. I need a bit more time to review.

tuxmachine commented:
@kamilmysliwiec Thanks for the heads up, I understand the size will require some time to review.

@lgabeskiria In the mean time, feel free to use my repository as a temporary fix. You can add it to your project using `npm install --save tuxmachine/graphql#feat/federation. I'm not inclined to publish it to NPM as I don't want to create competing graphql packages.

XBeg9 commented:
@tuxmachine, is there anything important fixed in your last batch of commits? Thanks

zgriesinger commented:
Thanks for your work on this @tuxmachine 

@kamilmysliwiec Do you have any updates on review/merge? Is there anything us invested in this issue can do to help?

robbyemmert commented:
Is there anything holding this PR up?

robbyemmert commented:
> Is there anything holding this PR up?

If it's just conflicts on the package.json, maybe I can help?

tuxmachine commented:
> Is there anything holding this PR up?

Nope it's just a large PR, takes time to review and maintainers are very busy.

kamilmysliwiec commented:
Thank you! Published as 6.6.0 :)

kauandotnet commented:
@tuxmachine, @marcus-sa  Thanks for yours contributions. 
Im excited to start working with federated schema on NestJS environment (Waiting docs merge).

@kamilmysliwiec Thanks for accepting, I believe you must be a very busy person. With this, NestJS becomes even more versatile, and its built-in modules are very helpful, making developers' lives a little easier.

tuxmachine commented:
@kamilmysliwiec Thanks for the review and merge! Feel free to ping me if issues come up

endurance commented:
How are we looking on the code-first implementation? Does it work? If not, what's the current blocker?

lookapanda commented:
@endurance I already have a service up and running on Amazon ECS using a code-first implementation! So it does semm to work! 🎉

Dragons0458 commented:
@endurance remember to use @Directive on your models.

For more information: https://github.com/MichalLytek/type-graphql/blob/master/examples/apollo-federation/inventory/product.ts

Only use the class Product and import all decorators from @nestjs/graphql

Regards!

kamilmysliwiec commented:
@tuxmachine would you like to create a PR with an update to the docs (code first way)? It works exactly the same as in e2e tests you wrote (`@Directive()` decorator etc)

tuxmachine commented:
@kamilmysliwiec Yeah no problem. I'll try to write the docs for code-first federation somewhere this month.

kamilmysliwiec commented:
Thank you @tuxmachine!

joemckie commented:
@lookapanda is there any chance you could share some basic code until the docs have been written? I'm attempting to migrate to a code-first federated server, but `GraphQLFederationFactory` is literally throwing an error that says "Code-first approach is not supported yet", so I'm just wondering how you did it 😄

tuxmachine commented:
@joemckie In rough lines: 

- You'll have to fork nest/graphql repo
- Delete the line throwing the error
- `npm install type-graphql@beta`
- Commit and push those changes to your fork
- In your project that wants to use schema-first: `npm install joemckie/graphql`

Look at the test files for some examples using code-first federation

joemckie commented:
@tuxmachine thanks! Will give that a try 😄

kamilmysliwiec commented:
@tuxmachine pinging in you here hoping you'll get a notification 😅 what's the best way to reach out to you? or better, can you message me on twitter/linkedin/via email (mail@kamilmysliwiec.com)?
